liquibase-core/src/main/java/liquibase/integration/servlet/GenericServletListener.java

```diff
@@ -0,0 +1,256 @@
+        final LiquibaseConfiguration liquibaseConfiguration = Scope.getCurrentScope().getSingleton(LiquibaseConfiguration.class);
+
+        setDataSource((String) liquibaseConfiguration.getCurrentConfiguredValue(null, null, LIQUIBASE_DATASOURCE).getValue());
+        if (getDataSource() == null) {
+            throw new RuntimeException("Cannot run Liquibase, " + LIQUIBASE_DATASOURCE + " is not set");
+        }
+
+        setChangeLogFile((String) liquibaseConfiguration.getCurrentConfiguredValue(null, null, LIQUIBASE_CHANGELOG).getValue());
+        if (getChangeLogFile() == null) {
+            throw new RuntimeException("Cannot run Liquibase, " + LIQUIBASE_CHANGELOG + " is not set");
+        }
+
+        setContexts((String) liquibaseConfiguration.getCurrentConfiguredValue(null, null, LIQUIBASE_CONTEXTS).getValue());
+        setLabels((String) liquibaseConfiguration.getCurrentConfiguredValue(null, null, LIQUIBASE_LABELS).getValue());
+        this.defaultSchema = StringUtil.trimToNull((String) liquibaseConfiguration.getCurrentConfiguredValue(null, null, LIQUIBASE_SCHEMA_DEFAULT).getValue());
+
+        Connection connection = null;
+        Database database = null;
+        Liquibase liquibase = null;
+        try {
+            DataSource dataSource = (DataSource) ic.lookup(this.dataSourceName);
+
+            connection = dataSource.getConnection();
+
+            Thread currentThread = Thread.currentThread();
+            ClassLoader contextClassLoader = currentThread.getContextClassLoader();
+            ResourceAccessor threadClFO = new ClassLoaderResourceAccessor(contextClassLoader);
+
+            ResourceAccessor clFO = new ClassLoaderResourceAccessor();
+            ResourceAccessor fsFO = new FileSystemResourceAccessor();
+
+
+            database = DatabaseFactory.getInstance().findCorrectDatabaseImplementation(new JdbcConnection(connection));
+            database.setDefaultSchemaName(getDefaultSchema());
+            liquibase = new Liquibase(getChangeLogFile(), new CompositeResourceAccessor(clFO, fsFO, threadClFO), database);
+
+            @SuppressWarnings("unchecked")
+            Enumeration<String> initParameters = servletContext.getInitParameterNames();
+            while (initParameters.hasMoreElements()) {
+                String name = initParameters.nextElement().trim();
+                if (name.startsWith(LIQUIBASE_PARAMETER + ".")) {
+                    liquibase.setChangeLogParameter(name.substring(LIQUIBASE_PARAMETER.length() + 1), liquibaseConfiguration.getCurrentConfiguredValue(null, null, name));
+                }
+            }
+
+            liquibase.update(new Contexts(getContexts()), new LabelExpression(getLabels()));
+            if (database instanceof DerbyDatabase) {
+                ((DerbyDatabase) database).setShutdownEmbeddedDerby(false);
```

Comment:
This feels a bit random. Why does Derby get a special treatment?

Reply:
It's existing code. Ideally we'd be handling it differently, but it's how it got implemented years ago.